RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc

```diff
@@ -10,44 +11,431 @@
+      (*lgc_map)[lay].extend_current(gp1.perp2());
+      (*lgc_map)[lay].extend_current(gp2.perp2());
+    }
+  }
+  if (lgc_map) {
+    (*lgc_map)[lay].add_current();
+  }
+  // Module information
+  auto p = det->position();
+  auto z = det->rotation().z();
+  auto x = det->rotation().x();
+  layer_info.register_module({{p.x(), p.y(), p.z()}, {z.x(), z.y(), z.z()}, {x.x(), x.y(), x.z()}, detid.rawId()});
+  // Set some layer parameters (repeatedly, would require hard-coding otherwise)
+  layer_info.set_subdet(detid.subdetId());
+  layer_info.set_is_pixel(detid.subdetId() <= 2);
+  layer_info.set_is_stereo(m_trackerTopo->isStereo(detid));
+}
+
+//==============================================================================
+
+// Ideally these functions would also:
+// 0. Setup LayerInfo data (which is now done in auto-generated code).
+//    Some data-members are a bit over specific, esp/ bools for CMS sub-detectors.
+// 1. Establish short module ids (now done in MkFitGeometry constructor).
+// 2. Store module normal and strip direction vectors
+// 3. ? Any other information ?
+// 4. Extract stereo coverage holes where they exist (TEC, all but last 3 double-layers).
+//
+// Plugin DumpMkFitGeometry.cc can then be used to export this for stand-alone.
+// Would also need to be picked up with tk-ntuple converter (to get module ids as
+// they will now be used as indices into module info vectors).
+//
+// An attempt at export cmsRun config is in python/dumpMkFitGeometry.py
+
+void MkFitGeometryESProducer::addPixBGeometry(mkfit::TrackerInfo &trk_info) {
+  for (auto &det : m_trackerGeom->detsPXB()) {
+    fillShapeAndPlacement(det, trk_info);
+  }
+}
+
+void MkFitGeometryESProducer::addPixEGeometry(mkfit::TrackerInfo &trk_info) {
+  for (auto &det : m_trackerGeom->detsPXF()) {
+    fillShapeAndPlacement(det, trk_info);
+  }
+}
+
+void MkFitGeometryESProducer::addTIBGeometry(mkfit::TrackerInfo &trk_info) {
+  for (auto &det : m_trackerGeom->detsTIB()) {
+    fillShapeAndPlacement(det, trk_info);
+  }
+}
+
+void MkFitGeometryESProducer::addTOBGeometry(mkfit::TrackerInfo &trk_info) {
+  for (auto &det : m_trackerGeom->detsTOB()) {
+    fillShapeAndPlacement(det, trk_info);
+  }
+}
+
+void MkFitGeometryESProducer::addTIDGeometry(mkfit::TrackerInfo &trk_info) {
+  for (auto &det : m_trackerGeom->detsTID()) {
+    fillShapeAndPlacement(det, trk_info);
+  }
+}
+
+void MkFitGeometryESProducer::addTECGeometry(mkfit::TrackerInfo &trk_info) {
+  // For TEC we also need to discover hole in radial extents.
+  layer_gap_map_t lgc_map;
+  for (auto &det : m_trackerGeom->detsTEC()) {
+    fillShapeAndPlacement(det, trk_info, &lgc_map);
+  }
+  // Now loop over the gap_collectors and see if there is a coverage gap.
+  gap_collector::interval itvl;
+  for (auto &[layer, gcol] : lgc_map) {
+    gcol.sqrt_elements();
+    if (gcol.find_gap(itvl, 0.5)) {
+      printf("TEC layer %d, gap: %f -> %f, width = %f\n", layer, itvl.x, itvl.y, itvl.y - itvl.x);
```

Comment:
printfs will need to be cleaned up, LogTrace/LogDebug is probably more appropriate, if still needed

Reply:
converted ... it's important top spot if a wrong gap (too small) gets picked up